File: C++/unique_length_3_palindromic_subsequences.cpp
//no. 1930
#include <string>
#include <set>
#include <map>
using namespace std;

class Solution {
public:
    int countPalindromicSubsequence(string s) {
        map <char, vector<int> > counter;
        for(int i = 0; i < s.length(); i++){
            counter[s[i]].push_back(i);
        }
        set<char> pan;
        int ans = 0;
        for(map<char, vector<int> >::iterator it = counter.begin(); it != counter.end(); it++){         
            pan.clear();
            char head = it->first;
            int start = it->second[0] + 1;
            int end = it->second.back();
            for (int i = start; i < end; i++){
                pan.insert(s[i]);
            }
            ans += pan.size();
        }
        return ans;
    }
};

File: go/unique_length_3_palindromic_subsequences.go
//no. 1930
func countPalindromicSubsequence(s string) int {
    counter := make(map[byte][]int)
    for i := 0; i < len(s); i++{
        counter[s[i]] = append(counter[s[i]], i)
    }
    var ans = 0
    pan := make(map[byte]bool)
    for k := range(counter){
        var start = counter[k][0] + 1
        var end = counter[k][len(counter[k]) - 1]
        for i := start; i < end; i++{
            pan[s[i]] = true
        }
        ans += len(pan)
        clear(pan)
    }
    return ans
}
